File: sources/CPP/include/Ers/Math/PathSegment.cpp
#include "PathSegment.h"
#include "Ers/Api.h"

namespace Ers
{
    PathSegmentType PathSegment::GetType() const
    {
        return static_cast<PathSegmentType>(ersAPIFunctionPointers.ERS_PathSegment_GetType((void*)this));
    }

    Vector3 PathSegment::From() const
    {
        return Vec3(
            ersAPIFunctionPointers.ERS_PathSegment_GetFromX((void*)this), ersAPIFunctionPointers.ERS_PathSegment_GetFromY((void*)this),
            ersAPIFunctionPointers.ERS_PathSegment_GetFromZ((void*)this));
    }

    Vector3 PathSegment::To() const
    {
        return Vec3(
            ersAPIFunctionPointers.ERS_PathSegment_GetToX((void*)this), ersAPIFunctionPointers.ERS_PathSegment_GetToY((void*)this),
            ersAPIFunctionPointers.ERS_PathSegment_GetToZ((void*)this));
    }

    Vector3 PathSegment::Center() const
    {
        return Vec3(
            ersAPIFunctionPointers.ERS_PathSegment_GetCenterX((void*)this), ersAPIFunctionPointers.ERS_PathSegment_GetCenterY((void*)this),
            ersAPIFunctionPointers.ERS_PathSegment_GetCenterZ((void*)this));
    }

    float PathSegment::GetRadius() const
    {
        return ersAPIFunctionPointers.ERS_PathSegment_GetRadius((void*)this);
    }

    float PathSegment::GetBeginAngle() const
    {
        return ersAPIFunctionPointers.ERS_PathSegment_GetBeginAngle((void*)this);
    }

    float PathSegment::GetEndAngle() const
    {
        return ersAPIFunctionPointers.ERS_PathSegment_GetEndAngle((void*)this);
    }

    float PathSegment::GetEndZ() const
    {
        return ersAPIFunctionPointers.ERS_PathSegment_GetEndZ((void*)this);
    }

    Vector3 PathSegment::GetPointAlongPath(float t) const
    {
        float result[3];
        ersAPIFunctionPointers.ERS_PathSegment_GetPointAlongPath((void*)this, t, result);
        return Vec3(result[0], result[1], result[2]);
    }

    Vector3 PathSegment::GetRotationAlongPath(float t) const
    {
        float result[3];
        ersAPIFunctionPointers.ERS_PathSegment_GetRotationAlongPath((void*)this, t, result);
        return Vec3(result[0], result[1], result[2]);
    }

    Vector3 PathSegment::GetP0() const
    {
        return Ers::Vec3(
            ersAPIFunctionPointers.ERS_PathSegment_GetP0X((void*)this), ersAPIFunctionPointers.ERS_PathSegment_GetP0Y((void*)this),
            ersAPIFunctionPointers.ERS_PathSegment_GetP0Z((void*)this));
    }

    Vector3 PathSegment::GetP1() const
    {
        return Ers::Vec3(
            ersAPIFunctionPointers.ERS_PathSegment_GetP1X((void*)this), ersAPIFunctionPointers.ERS_PathSegment_GetP1Y((void*)this),
            ersAPIFunctionPointers.ERS_PathSegment_GetP1Z((void*)this));
    }

    Vector3 PathSegment::GetP2() const
    {
        return Ers::Vec3(
            ersAPIFunctionPointers.ERS_PathSegment_GetP2X((void*)this), ersAPIFunctionPointers.ERS_PathSegment_GetP2Y((void*)this),
            ersAPIFunctionPointers.ERS_PathSegment_GetP2Z((void*)this));
    }

    Vector3 PathSegment::GetP3() const
    {
        return Ers::Vec3(
            ersAPIFunctionPointers.ERS_PathSegment_GetP3X((void*)this), ersAPIFunctionPointers.ERS_PathSegment_GetP3Y((void*)this),
            ersAPIFunctionPointers.ERS_PathSegment_GetP3Z((void*)this));
    }

    void* PathSegment::Data()
    {
        return this;
    }

    const void* const PathSegment::Data() const
    {
        return this;
    }
} // namespace Ers

File: sources/CSharp/src/Ers/Math/PathSegment.cs
using System;
using System.Numerics;
using Ers.Engine;

namespace Ers
{
    public enum PathSegmentType
    {
        Straight,
        Helical,
        CubicBezier,
    }

    public ref struct PathSegment
    {
        /// <summary>
        /// Native pointer to the core instance.
        /// </summary>
        public readonly IntPtr CorePtr;

        public PathSegment(IntPtr corePtr) => CorePtr = corePtr;

        public PathSegmentType Type
        {
            get => (PathSegmentType)ErsEngine.ERS_PathSegment_GetType(CorePtr);
        }

        public Vector3 From
        {
            get => new Vector3(
                ErsEngine.ERS_PathSegment_GetFromX(CorePtr),
                ErsEngine.ERS_PathSegment_GetFromY(CorePtr),
                ErsEngine.ERS_PathSegment_GetFromZ(CorePtr));
        }

        public Vector3 To
        {
            get => new Vector3(
                ErsEngine.ERS_PathSegment_GetToX(CorePtr),
                ErsEngine.ERS_PathSegment_GetToY(CorePtr),
                ErsEngine.ERS_PathSegment_GetToZ(CorePtr));
        }

        public Vector3 Center
        {
            get => new Vector3(
                ErsEngine.ERS_PathSegment_GetCenterX(CorePtr),
                ErsEngine.ERS_PathSegment_GetCenterY(CorePtr),
                ErsEngine.ERS_PathSegment_GetCenterZ(CorePtr));
        }

        public float Radius => ErsEngine.ERS_PathSegment_GetRadius(CorePtr);

        public float BeginAngle => ErsEngine.ERS_PathSegment_GetBeginAngle(CorePtr);

        public float EndAngle => ErsEngine.ERS_PathSegment_GetEndAngle(CorePtr);

        public float EndZ => ErsEngine.ERS_PathSegment_GetEndZ(CorePtr);

        public Vector3 P0
        {
            get => new Vector3(
                ErsEngine.ERS_PathSegment_GetP0X(CorePtr),
                ErsEngine.ERS_PathSegment_GetP0Y(CorePtr),
                ErsEngine.ERS_PathSegment_GetP0Z(CorePtr));
        }

        public Vector3 P1
        {
            get => new Vector3(
                ErsEngine.ERS_PathSegment_GetP1X(CorePtr),
                ErsEngine.ERS_PathSegment_GetP1Y(CorePtr),
                ErsEngine.ERS_PathSegment_GetP1Z(CorePtr));
        }

        public Vector3 P2
        {
            get => new Vector3(
                ErsEngine.ERS_PathSegment_GetP2X(CorePtr),
                ErsEngine.ERS_PathSegment_GetP2Y(CorePtr),
                ErsEngine.ERS_PathSegment_GetP2Z(CorePtr));
        }

        public Vector3 P3
        {
            get => new Vector3(
                ErsEngine.ERS_PathSegment_GetP3X(CorePtr),
                ErsEngine.ERS_PathSegment_GetP3Y(CorePtr),
                ErsEngine.ERS_PathSegment_GetP3Z(CorePtr));
        }

        public Vector3 GetPointAlongPath(float t)
        {
            unsafe
            {
                float* result = stackalloc float[3];
                ErsEngine.ERS_PathSegment_GetPointAlongPath(CorePtr, t, (IntPtr)result);
                return new Vector3(result[0], result[1], result[2]);
            }
        }

        public Vector3 GetRotationAlongPath(float t)
        {
            unsafe
            {
                float* result = stackalloc float[3];
                ErsEngine.ERS_PathSegment_GetRotationAlongPath(CorePtr, t, (IntPtr)result);
                return new Vector3(result[0], result[1], result[2]);
            }
        }
    }
}
